client-core/src/wallet/syncer.rs

```diff
@@ -0,0 +1,546 @@
+        );
+        Ok(Self::with_config(
+            SyncerConfig {
+                storage: config.storage,
+                client: config.client,
+                enable_fast_forward: config.enable_fast_forward,
+                batch_size: config.batch_size,
+            },
+            decryptor,
+            progress_reporter,
+            name,
+            passphrase,
+        ))
+    }
+}
+
+/// Sync wallet state from blocks.
+struct WalletSyncerImpl<'a, S: SecureStorage, C: Client, D: TxDecryptor> {
+    env: &'a WalletSyncer<S, C, D>,
+
+    // cached state
+    wallet: Wallet,
+    sync_state: SyncState,
+    wallet_state: WalletState,
+}
+
+impl<'a, S: SecureStorage, C: Client, D: TxDecryptor> WalletSyncerImpl<'a, S, C, D> {
+    fn new(env: &'a WalletSyncer<S, C, D>) -> Result<Self> {
+        let wallet = service::load_wallet(&env.storage, &env.name, &env.passphrase)?
+            .err_kind(ErrorKind::InvalidInput, || {
+                format!("wallet not found: {}", env.name)
+            })?;
+
+        let sync_state = service::load_sync_state(&env.storage, &env.name)?;
+        let sync_state = if let Some(sync_state) = sync_state {
+            sync_state
+        } else {
+            SyncState::genesis(env.client.genesis()?.validators)
+        };
+
+        let wallet_state = service::load_wallet_state(&env.storage, &env.name, &env.passphrase)?
+            .unwrap_or_default();
+
+        Ok(Self {
+            env,
+            wallet,
+            sync_state,
+            wallet_state,
+        })
+    }
+
+    fn init_progress(&self, height: u64) {
+        if let Some(ref sender) = &self.env.progress_reporter {
+            let _ = sender.send(ProgressReport::Init {
+                wallet_name: self.env.name.clone(),
+                start_block_height: self.sync_state.last_block_height,
+                finish_block_height: height,
+            });
+        }
+    }
+
+    fn update_progress(&self, height: u64) {
+        if let Some(ref sender) = &self.env.progress_reporter {
+            let _ = sender.send(ProgressReport::Update {
+                wallet_name: self.env.name.clone(),
+                current_block_height: height,
+            });
+        }
+    }
+
+    fn save(&mut self, memento: &WalletStateMemento) -> Result<()> {
+        service::save_sync_state(&self.env.storage, &self.env.name, &self.sync_state)?;
+        self.wallet_state = service::modify_wallet_state(
+            &self.env.storage,
+            &self.env.name,
+            &self.env.passphrase,
+            |state| state.apply_memento(memento),
+        )?;
+        Ok(())
+    }
+
+    fn handle_batch(&mut self, blocks: NonEmpty<BlockHeader>) -> Result<()> {
+        let enclave_txids = blocks
+            .iter()
+            .flat_map(|block| block.enclave_transaction_ids.iter().copied())
+            .collect::<Vec<_>>();
+        let enclave_txs = self.env.decryptor.decrypt_tx(&enclave_txids)?;
```

Comment:
Have you tested this? What if `enclave_txids` is an empty vector?

Reply:
Yes, it passes `cargo test` and integration tests locally.
The result of `decrypt_tx` depends on the obfuscation implementation, maybe we should check emptiness to avoid the call to `decrypt_tx`.

Reply:
Currently, `BlockGenerator` doesn't support tx mocking yet, after that implemented, we can test all the syncer behavior in unit testing.